cpp/src/kpca/kpca.cuh

```diff
@@ -0,0 +1,180 @@
+ */
+template <typename value_t, typename enum_solver = solver>
+void kpcaFit(const raft::handle_t &handle, value_t *input, value_t *alphas,
+             value_t *lambdas, const paramsKPCA &prms, cudaStream_t stream) {
+  auto cublas_handle = handle.get_cublas_handle();
+  auto allocator = handle.get_device_allocator();
+  ASSERT(prms.n_cols > 1,
+         "Parameter n_cols: number of columns cannot be less than two");
+  ASSERT(prms.n_rows > 1,
+         "Parameter n_rows: number of rows cannot be less than two");
+  ASSERT(
+    prms.n_components > 0,
+    "Parameter n_components: number of components cannot be less than one");
+
+  int n_components = prms.n_components;
+  if (n_components > prms.n_rows) n_components = prms.n_rows;
+
+  Matrix::GramMatrixBase<value_t> *kernel =
+    Matrix::KernelFactory<value_t>::create(prms.kernel, cublas_handle);
+  
+  value_t *kernel_mat;
+  raft::allocate(kernel_mat, prms.n_rows * prms.n_rows);
+  kernel->evaluate(input, prms.n_rows, prms.n_cols, input, prms.n_rows,
+         kernel_mat, false, stream, prms.n_rows, prms.n_rows, prms.n_rows);
+
+  //  create centering matrix (I - 1/nrows)
+  value_t * i_mat;
+  raft::allocate(i_mat, prms.n_rows * prms.n_rows);
+
+  value_t * diag;
+  raft::allocate(diag, prms.n_rows);
+  auto thrust_policy = rmm::exec_policy(stream);
+  thrust::fill(thrust_policy, diag, diag + prms.n_rows, 1.0f);
+
+
+  //  create centering matrix 
+  value_t * centering_mat;
+  raft::allocate(centering_mat, prms.n_rows * prms.n_rows);
+  raft::matrix::initializeDiagonalMatrix(diag, i_mat, prms.n_rows, prms.n_rows, stream);
+  value_t inv_n_rows = 1.0/prms.n_rows;
+  raft::linalg::subtractScalar(centering_mat, i_mat, inv_n_rows, prms.n_rows * prms.n_rows, stream);
+
+  //  center the kernel matrix: K' = (I - 1/nrows) K (I - 1/nrows)
+  value_t * temp_mat;
+  raft::allocate(temp_mat, prms.n_rows * prms.n_rows);
+  value_t alpha = 1.0;
+  value_t beta = 0.0;
+  
+  raft::linalg::gemm(handle, centering_mat, prms.n_rows, prms.n_rows
+              , kernel_mat, temp_mat, prms.n_rows, prms.n_rows
+              , CUBLAS_OP_N, CUBLAS_OP_N, alpha, beta, stream);
+
+  value_t * centered_kernel;
+  raft::allocate(centered_kernel, prms.n_rows * prms.n_rows);
+
+  raft::linalg::gemm(handle, temp_mat, prms.n_rows, prms.n_rows
+              , centering_mat, centered_kernel, prms.n_rows, prms.n_rows
+              , CUBLAS_OP_N, CUBLAS_OP_N, alpha, beta, stream);
+
+  //  either Jacobi (iterative power method) or DnC eigendecomp
+  if (prms.algorithm == enum_solver::COV_EIG_JACOBI) {
+    raft::linalg::eigJacobi(handle, centered_kernel, prms.n_rows, prms.n_rows, alphas,
+                            lambdas, stream, (value_t)prms.tol,
+                            prms.n_iterations);
+  } else {
+    raft::linalg::eigDC(handle, centered_kernel, prms.n_rows, prms.n_rows, alphas,
+                            lambdas, stream);
+  }
+  raft::matrix::colReverse(alphas, prms.n_rows, prms.n_rows, stream);
```

Comment:
Can you explain why these reverses are needed?

Reply:
@tjohanne was getting his eigenvectors and eigenvalues in arbitrary order, and we noticed that this is how PCA.cuh handles the unsorted outputs.  I will look into it deeper and verify that it is the best way to handle this.

Reply:
According to the cuSOLVER documentation for cusolverDnsyevd it returns "The eigenvalue values of A, in ascending order ie, sorted so that W(i) <= W(i+1)." We wanted the eigenvalues in descending order. 
The reverse operations are also done in [tsvd.cuh](https://github.com/rapidsai/cuml/blob/4d069915b9d12fa8b301d1c5e4bb4abe5198d1a4/cpp/src/tsvd/tsvd.cuh#L100)